x/programs/rust/wasmlanche-sdk/src/state.rs

```diff
@@ -68,30 +96,47 @@ impl State {
     /// the host fails to read the key and value.
     /// # Panics
     /// Panics if the value cannot be converted from i32 to usize.
-    pub fn get<T, K>(&self, key: K) -> Result<T, Error>
+    pub fn get<V>(&mut self, key: K) -> Result<V, Error>
     where
-        K: Into<Key>,
-        T: BorshDeserialize,
+        V: BorshDeserialize,
     {
-        let val_ptr = unsafe { host::get_bytes(&self.program, &key.into())? };
-        if val_ptr < 0 {
-            return Err(Error::Read);
-        }
+        let val_bytes = if let Some(val) = self.cache.get(&key) {
+            val.clone()
+        } else {
+            let val_ptr = unsafe { host::get_bytes(&self.program, &key.clone().into())? };
+            if val_ptr < 0 {
+                return Err(Error::Read);
+            }
+
+            // Wrap in OK for now, change from_raw_ptr to return Result
+            into_bytes(val_ptr)
+        };
 
-        // Wrap in OK for now, change from_raw_ptr to return Result
-        unsafe { from_host_ptr(val_ptr) }
+        self.cache.insert(key, val_bytes.clone());
+
+        from_slice::<V>(&val_bytes).map_err(|_| StateError::Deserialization)
```

Comment:
Did you try using the `HashMap::entry` method? It should enable you do get `val_bytes` without cloning anywhere.

Reply:
I did tried, but the entry API doesn't provide any convenient method other than chaining `and_modify` with `or_insert`, moreover the value of the `HashMap` is owned so I'm not sure if we can get around that.

Reply:
By permutating the line that calls `from_slice` with the one that inserts in the map, I could get around that clone.

Reply:
Not sure what you mean by "permutating".

I think it should look something more like this:
```rust
    pub fn get<V>(&mut self, key: K) -> Result<Option<V>, Error>
    where
        V: BorshDeserialize,
    {
        if let Some(val) = self.cache.get(&key) {
            return from_slice(&val)
                .map_err(|_| Error::Deserialization)
                .map(Some);
        }

        let host_ptr = unsafe {
            // cloning the key here isn't ideal, but we should be able to fix this later
            host::get_bytes(&self.program, &key.clone().into()).map_err(|_| Error::Read)?
        };

        // I'm not 100% that this is how the host handles things when a key doesn't exist
        if host_ptr == 0 {
            return Ok(None);
        }

        let bytes = into_bytes(host_ptr);
        let val = self.cache.entry(key).or_insert(bytes);

        from_slice(&val)
            .map_err(|_| Error::Deserialization)
            .map(Some)
    }
```